### DISCUSSION TITLE
Hosting CoreWCF on Docker instead of IIS

### DISCUSSION THREAD

**Discussion by VirReal1** [CoreWCF/CoreWCF#1468] [Q&A]
Hello again. I'm trying to host my application in docker instead of IIS but have some issues regarding the generated wsdl file. In IIS, the wsdl file endpoints would automatically be the site's bindings, but for docker, I'm using Kestrel and the generated wsdl file is a little problematic. 
My applications both have None and Transport bindings. When I request the wsdl file using the https scheme, the wsdl file generates https endpoints correctly, but the http endpoint defaults to localhost. If I request it using http, the same thing happens to the https. This is probably because I'm using docker with the Kestrel web server.
I tried setting Urls for Kestrel by using ASPNETCORE_URLS environment variable but that did not work. I do not want to set BaseAddress because I need to be able to deploy a single build to many places with different domain names.
Any help will be much appreciated.

**Comment by mconnew** [comment]
If you add `services.AddSingleton<IServiceBehavior, UseRequestHeadersForMetadataAddressBehavior>();` it will pull the hostname from the request `Host` header that the client sends. Set everything up so it works with the hostname localhost, which might mean setting up the listen addresses on the metadata behavior like this:
```csharp
app.UseServiceModel(serviceBuilder =>
{
    serviceBuilder.AddService<TService>();
    serviceBuilder.AddServiceEndpoint<TService, TContract>(binding, url);
});
var serviceMetadataBehavior = app.ApplicationServices.GetRequiredService<CoreWCF.Description.ServiceMetadataBehavior>();
serviceMetadataBehavior.HttpGetEnabled = true;
serviceMetadataBehavior.HttpsGetEnabled = true;
serviceMetadataBehavior.HttpGetUrl = new Uri($"http://localhost/myServicePath/MyService.svc");
serviceMetadataBehavior.HttpsGetUrl = new Uri($"https://localhost/myServicePath/MyService.svc");
```
When you use `UseRequestHeadersForMetadataAddressBehavior`, when the WSDL is generated, as the url endpoint elements are being emitted, it replaces the hostname (and port number if applicable) with the values from the Host header.

**Comment by VirReal1** [comment]
I have already injected the UseRequestHeadersForMetadataAddressBehavior class but the problem continues. WSDL generates 2 sets of url's one for http and one https. If the wsdl request comes from http, wsdl uses the Host header value for http bindings but https bindings use localhost. If request comes from https, this time the http bindings are localhost. UseRequestHeadersForMetadataAddressBehavior only overrides the binding with the matching http scheme. I do not have this problem in IIS because I can add http and https bindings and it uses corresponding bindings to generate the http and https bindings.

**Reply by mconnew** [reply]
That sounds like a bug, it should replace all of them.

**Reply by mconnew** [reply]
I'll try to put together an alternative workaround using IWsdlExporter extensibility point. You can edit anything in the wsdl with that. But I'm not sure if that will work, I'll have to see.

**Reply by VirReal1** [reply]
I do not know if it helps but the UseRequestHeadersForMetadataAddressBehavior class only triggers once when generating the wsdl and it is for the matching scheme. For other scheme, it is not triggered.

**Reply by VirReal1** [reply]
Hello @mconnew,

After delving deeper into WCF, I realized that the behavior I encountered is actually by design, and it's consistent with how .NET Framework's WCF operates.

In my setup, there are two URLs associated with the WSDL: one for HTTP and one for HTTPS. When I request the WSDL over HTTP, it only replaces the HTTP URLs in the WSDL with the values from the Host header, leaving the HTTPS URLs unchanged. I now understand this is the intended behavior, as .NET Framework WCF follows the same approach—it only updates the URLs for the corresponding scheme, leaving the others as they are.

Looking back, I could have phrased my original issue more clearly. You may have thought it wasn't replacing all the URLs for the relevant scheme, but that wasn't the case.

Thanks for your help, and I hope you have a great day!
